Simplify custom templates management -> only a single `--template` argument

guillemcordoba opened the pull request:
Closes https://github.com/holochain/scaffolding/issues/213.

My approach with this PR has been a bit drastic maybe, in that I've completely dropped the support for the `--templates-url` argument since I don't have a way to have good developer experience with it. Instead, the management of custom templates is refactored and simplified: we go from 3 possible arguments `--template`, `--templates-url` and `--templates-path` to only one argument `--template`, that either specifies the built-in template to use, or a path to the custom template to use.

This is an example of a custom template already pointing to this branch:

https://github.com/holochain-open-dev/templates

c12i reviewed (commented):
I like the idea of combining the template related subcommands into one `--template`.

However, the removal of `--templates-url` restricts custom templates to Nix, a drastic change indeed. While I'm open to this shift, it's crucial to document it explicitly. I propose updating [this guide](https://docs.rs/holochain_scaffolding_cli/0.2000.0/holochain_scaffolding_cli/#how-to-create-a-custom-template) to guide users on creating custom templates using the nix wrapper approach.

Additionally, the global nature of `--template` introduces a potential issue, specifically for non-nix users. For instance, an initially scaffolded `svelte` web app could (un)intentionally have `hc-scaffold --template vue entry-type` commands run against it. I haven't come across code handling such scenarios, and I think it might be worth addressing.

guillemcordoba commented:
I agree with your points. I've added instructions on how to set up the custom template the nix way, and added a nix flake template for a custom scaffolding tool template so that it's easy to get started on that.

Regarding the `--template` argument, I see what you mean. How would you like to move forward? I would be happy to change it to a `CUSTOM_TEMPLATES_PATH` env var, or some other solution that you would see fit, my only requirement is that it can be set at every command somehow. I don't have the bandwidth right now to contribute a more complex solution on a manifest or similar, but I'd be happy to help you out if you want to take this branch forward and add that kind of custom template management on top of this PR.

c12i commented:
Yes an env var would be good idea, when a user selects the template they would like to scaffold with when running `hc-scaffold web-app`, we save this value to `CUSTOM_TEMPLATES_PATH`. Subsequent commands that explicitly have `--template` flag can have a check to ensure that the selected template/ path conforms with the original template/path set when the happ was first scaffolded

guillemcordoba commented on flake.nix:
```diff
@@ -37,6 +64,9 @@
         };
+
+        # TODO: Expose the scaffolding tool CLI as the main package for this crate
+        # packages.default = inputs'.holochain.packages.hc-scaffold;
```
Mmm we _could_, but it wouldn't actually be exposing the package coming from the source code of this repository, it would be exposing the one that's coming from the holochain main repository, defined [here](https://github.com/holochain/holochain/blob/392bdfd729fb6ce50f78f9e7f1c757dc392675f4/versions/weekly/flake.nix#L22) and built [here](https://github.com/holochain/holochain/blob/develop/nix/modules/scaffolding.nix). I think the build logic for scaffolding should be declared in this repository, and the main holochain repo should just point to it instead declaring it in the main holochain repository. What do you think @steveej ?

guillemcordoba commented:
@c12i I don't think I'm following you with the suggestion of the `CUSTOM_TEMPLATE_PATH` env var. Would that env var only be set in the first `web-app` command, and saved somewhere? My only requirement is to be able to set the custom template path _at every command_, that's the only thing I need to be able to wrap the custom templates. So when I proposed the env var solution I was thinking along the lines of:

```bash
CUSTOM_TEMPLATE_PATH=../my/custom/template hc scaffold web-app
CUSTOM_TEMPLATE_PATH=../my/custom/template hc scaffold dna
CUSTOM_TEMPLATE_PATH=../my/custom/template hc scaffold zome
CUSTOM_TEMPLATE_PATH=../my/custom/template hc scaffold entry-type
```

And so the `--templates` argument would not be necessary at all, except for the `web-app` command.

I don't think that matches your proposal does it? What do you think of this way of doing it?

c12i commented:
@guillemcordoba sorry, I thought your suggestion was to solve the potential issue of using `--template vue` and `--template svelte` on the same project

guillemcordoba commented:
Oh okey... How do we suggest to move forward with this then?

c12i commented:
An approach similar to what what I suggested here could fix this

https://github.com/holochain/scaffolding/pull/229#issuecomment-1978731537

guillemcordoba commented:
Okey... What I was trying to convey is that I didn't understand what you meant exactly, and how would that approach work. Could you be a bit more specific about it or explain it in other words?

c12i commented:
When a template is chosen when `hc-scaffold web-app` is run, we could store the selected template, either in memory or as an env var, so that when subsequent `hc-scaffold` commands are called, specifically those which explicitly have the `--template` flag passed, we can have a check in place to ensure that the template passed still matches the original template the project was scaffolded with, hope my explanation is clear

c12i commented:
> That strategy is not resilitent to system reboots, and the env var would not be shared all the contributors of the repository so each one of them will end up with a different custom template, no?

You are right, this would not hold well. My second suggestion would be keeping a dotfile in the scaffolded project that contains this information, and this file would need to be committed in to git. Not really the most elegant solution but more resilient. What do you think?

c12i commented:
https://github.com/holochain/scaffolding/pull/229#issuecomment-1979461576

What I suggest here seems to go back to the conversation around the need for keeping a manifest file of some sort in the scaffolded project, to store this kind of information. I am happy to take on this branch and implement this kind of solution

guillemcordoba commented:
Yeah that makes sense! Thanks, I'll go about my other things from now on but can help out in this branch if you have a specific request!

c12i commented:
@guillemcordoba I added functionality to write and read scaffold config in the hApp's `package.json` file. Works as expected for hApps scaffolded with a path to a template passed as the `--template`.

I have not been able to successfuly test whether this still works as expected when scaffolding with a custom template via `nix run github:c12i/templates#hc-scaffold-app-template -- web-app`. I keep running into this error:

```shell
error:
       … while calling the 'derivationStrict' builtin

         at /builtin/derivation.nix:9:12: (source not available)

       … while evaluating derivation 'hc-scaffold'
         whose name attribute is located at /nix/store/bjvqq8c79dbi59g7xzcc6lhl0f19m3d7-source/pkgs/stdenv/generic/make-derivation.nix:353:7

       … while evaluating attribute 'buildCommand' of derivation 'hc-scaffold'

         at /nix/store/bjvqq8c79dbi59g7xzcc6lhl0f19m3d7-source/pkgs/build-support/trivial-builders/default.nix:98:16:

           97|         enableParallelBuilding = true;
           98|         inherit buildCommand name;
             |                ^
           99|         passAsFile = [ "buildCommand" ]

       error: cannot fetch input 'path:.?lastModified=0&narHash=sha256-VZvcESfdHPlP9t5mkGudilN2SO6i6CJm95Q/NMe/oYg%3D' because it uses a relative path
```

guillemcordoba commented:
Hey thanks for continuing this work. Hum... I'm not getting that error right now if I run that command. This is quite weird to me, am I missing anything that could be different in your environment?

c12i commented:
Likely, I have tested on MacOS and Debian, still the same error. Perhaps it might be worth verifying from your end specifically how the `customTemplatePath` is getting resolved, my guess is that this could be the issue

c12i commented:
The first link you shared was my attempt at trying to fix the issue/
I am running nix version `2.18.2` on both OSs. On Debian in fact, it was a fresh installation of nix.

guillemcordoba commented:
Hi @c12i I was able to reproduce the issue on another computer... Really weird. Anyway I pushed a solution to the issue, although I'm not bumping into another problem: 

```
guillem@guillems-thinkpad:~/projects/holochain/core/scaffoscaffolding$ nix run .#custom-template -- help
warning: Git tree '/home/guillem/projects/holochain/core/scaffolding' is dirty
error: Found argument '--template' which wasn't expected, or isn't valid in this context

USAGE:
    hc-scaffold <SUBCOMMAND>

For more information try --help
```

This is because the scaffolding tool that's the output of this flake is not what's on this "nixify" branch, but rather whatever is [pointed to by holochain](https://github.com/holochain/holochain/blob/ddeb601b7276db57990a76defbb2966229cc6f3d/versions/weekly/flake.nix#L22), so "holochain-weekly". We won't be able to properly test it until it gets released from that branch and updated to in the flake from the holochain repository. I think this is a problem with the way that the nix infrastructure is set up, and that it's actually this repository the one that should have [this code](https://github.com/holochain/holochain/blob/develop/nix/modules/scaffolding.nix).

I'm fine to merge this as is and create another issue for @steveej to indicate the best path of action regarding the nix infrastructure.

c12i commented:
> but rather whatever is [pointed to by holochain](https://github.com/holochain/holochain/blob/ddeb601b7276db57990a76defbb2966229cc6f3d/versions/weekly/flake.nix#L22), so "holochain-weekly"

I see, so merging this change into `holochain-weekly` should at least allow us to test this out then. Or does [this code](https://github.com/holochain/holochain/blob/develop/nix/modules/scaffolding.nix) also need to be in the scaffolding repo for this to work?

guillemcordoba commented:
No it doesn't need moving for us to test it. I think it _should_ be moved here to allow direct testing of branches from this repository, but it's not a necessity right now. So I'd go ahead and merge this for now.
